### PR TITLE
Speculative processing

### PR DESCRIPTION
**Pull request by mbhaskar**
# Description
This PR adds support for speculative processing on top of end-to-end timeout to improve availability within timeout

Introduces Threshold based speculative execution
==
Only for Reads / Queries
If we don't get a successful response (either delayed response or error response) within the threshold value, we go to all preferred regions in parallel. 
        - We can try one region at a time, with some linear timeline.
        - Whichever responds first, we will return, and cancel all the other ongoing requests. 
Benefits:
    1. Better chances of getting a response from either of those regions.
    2. This provides **Maximum** availability at RUs / compute tradeoff (increased cost)

```mermaid
flowchart TD
    A[Request] -->B(Base Request with timeout policy)
    B --> Threshold{timespent < speculativeThreshold}
    Threshold -->| Yes | C{success <= EndToEndTimeout}
    Threshold --> | No | RemoteProcessing[Start Parallel remote requests to other regions]
    C -->|Yes| D[Return result]
    C -->|No response| E[Cancel and timeout]
    C --> |error response| F[Retry Flow]
    F --> G{timespent < EndToEndTimeout}
    G --> |No| E
    G --> |Yes| F
```

If speculative retries are enabled, based on the probability threshold Speculative retries may increase RU costs

When we have sparse data, the algo adds more variance around mean and this may increase the exploration even though the lower mean value is selected until the perf guarantee is increased. (The algo has proofs that this keeps cost under consideration and does only necessary exploration)

Current algo is tuned for 10sample requests. It can be tuned based on time/ number of requests and can be adjusted accordingly

Initial test resuls
<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:dt="uuid:C2F41010-65B3-11d1-A29F-00AA00C14882"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=OneNote.File>
<meta name=Generator content="Microsoft OneNote 15">
</head>

<body lang=en-US style='font-family:Calibri;font-size:11.0pt'>

<p style='margin:0in;font-family:Calibri;font-size:11.0pt' lang=en-CA>Tests:</p>

<p style='margin:0in;font-family:Calibri;font-size:11.0pt' lang=en-CA>&nbsp;</p>

<p style='margin:0in;font-family:Calibri;font-size:11.0pt' lang=en-CA>Read
document 100 times</p>

<p style='margin:0in;font-family:Calibri;font-size:11.0pt' lang=en-CA>30s of
delay faults on the local region</p>

<div style='direction:ltr'>

  | P99 in   ms | P95 in   ms | P75 in   ms |  
-- | -- | -- | -- | --
No   Speculation | 2078.12109375 | 2028.05859375 | 2016.05859375 |  
Threshold   based Speculation | 1134.046875 | 616.015625 | 298.0 |  

How to enable:

Following environment variables enable speculation when end to end timeout policy is specified
 Integer value specifying the speculation type

 0 - No speculation
 1 - Threshold based speculation

**COSMOS_SPECULATION_TYPE**

The threshold int representing milli seconds
**COSMOS_SPECULATION_THRESHOLD**

The threshold step int representing milli seconds. This sets when speculation on subsequent regions is triggered
**COSMOS_SPECULATION_THRESHOLD_STEP**

example: Three regions r1, r2,r3 in preferred regions
Speculation threshold: 500
Specialization step: 100

At time t1 request to r1 is made
If there is no response in 500ms, request to r2 is made
If there is no response at 500+100ms, request to r3 is made

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/pom.xml [hunk 1]
```diff
@@ -89,6 +89,12 @@ Licensed under the MIT License.
       <version>1.7.36</version> <!-- {x-version-update;org.slf4j:slf4j-api;external_dependency} -->
     </dependency>
 
+    <dependency>
```
Comment (mbhaskar) [inline sdk/cosmos/azure-cosmos/pom.xml:92]: This will be shaded in the next commit

File: sdk/cosmos/azure-cosmos/src/main/java/module-info.java [hunk 2]
```diff
@@ -27,8 +27,9 @@
     //  This is only required by guava shaded libraries
     requires java.logging;
 	requires HdrHistogram;
+    requires commons.math3;
```
Comment (mbhaskar) [inline sdk/cosmos/azure-cosmos/src/main/java/module-info.java:30]: not required once the library is shaded

Comment (mbhaskar) [reply sdk/cosmos/azure-cosmos/src/main/java/module-info.java]: Removed

File: sdk/cosmos/azure-cosmos/pom.xml (hunk 1 above)
Comment (mbhaskar) [reply sdk/cosmos/azure-cosmos/pom.xml]: math3 is shaded so this is resolved

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java [hunk 3]
```diff
@@ -20,6 +20,18 @@
 
 public class Configs {
     private static final Logger logger = LoggerFactory.getLogger(Configs.class);
+    public static final String SPECULATION_ENABLED = "COSMOS_SPECULATION_ENABLED";
```
Comment (mbhaskar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java:23]: This should be removed

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/HttpConstants.java [hunk 4]
```diff
@@ -371,6 +371,10 @@ public static class StatusCodes {
 
         public static final int SERVICE_UNAVAILABLE = 503;
         public static final int INTERNAL_SERVER_ERROR = 500;
+        /**
+         * Used by {@link RequestCancelledException} to indicate a client side request cancellation
+         */
+        public static final int REQUEST_CANCELLED = 499;
     }
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/HttpConstants.java:378]: No 499 - like discussed in end2end timeout PR

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/DocumentServiceRequestContext.java [hunk 5]
```diff
@@ -124,7 +126,17 @@ public DocumentServiceRequestContext clone() {
         context.resourcePhysicalAddress = this.resourcePhysicalAddress;
         context.throughputControlCycleId = this.throughputControlCycleId;
         context.replicaAddressValidationEnabled = this.replicaAddressValidationEnabled;
+        context.e2EOperationRetryPolicyConfig = this.e2EOperationRetryPolicyConfig;
+        context.effectivePartitionKey = this.effectivePartitionKey;
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/DocumentServiceRequestContext.java:130]: duplicate - done in line 123 already

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java [hunk 6]
```diff
@@ -20,6 +20,18 @@
...
+
+    /**
+     * Integer value specifying the speculation type
+     * <pre>
+     * 0 - No speculation
+     * 1 - Threshold based speculation
+     * 2 - Thompson Sampling based speculation
+     * </pre>
+     */
+    public static final String SPECULATION_TYPE = "COSMOS_SPECULATION_TYPE";
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java:33]: I am wondering whether we should provide both choices (threshold based and Thompson sampling). each choices have its strong and weakness, when we try to improve those areas, eventually it may drive into one hybrid solution. 
Should we start with one choice and keep making improvement on?

Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java:33]: And another question: should we give choice 0

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java [hunk 7]
```diff
@@ -85,6 +94,19 @@ public ReplicatedResourceClient(
...
         this.enableReadRequestsFallback = enableReadRequestsFallback;
+
+        switch (Configs.getSpeculationType()) {
+            case SpeculativeProcessor.THRESHOLD_BASED:
+                this.speculativeProcessor = new ThresholdBasedSpeculation();
+                break;
+            case SpeculativeProcessor.THOMPSON_SAMPLING_BASED:
+                this.speculativeProcessor = new ThompsonSamplingBasedSpeculation(this.transportClient
+                    .getGlobalEndpointManager()
+                    .getAvailableReadEndpoints());
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java:105]: use readEndpoints instead?

I think the difference between availableEndpoints vs readEndpoints is:
availableEndpoints -> the regions configured for the account
readEndpoint -> the regions based on the PreferredRegions and some other criteria.

For speculative processing, we probably should still honor the preferredRegion etc config

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java [hunk 8]
```diff
@@ -113,8 +135,12 @@ public Mono<StoreResponse> invokeAsync(RxDocumentServiceRequest request,
             documentServiceRequest.getHeaders().put(HttpConstants.HttpHeaders.REMAINING_TIME_IN_MS_ON_CLIENT_REQUEST,
                     Long.toString(forceRefreshAndTimeout.getValue2().toMillis()));
 
-            return invokeAsync(request, new TimeoutHelper(forceRefreshAndTimeout.getValue2()),
-                forceRefreshAndTimeout.getValue1(), forceRefreshAndTimeout.getValue0());
+            if (shouldSpeculate(request)){
+                logger.info("Speculating request {}", request);
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java:139]: should this be debug level?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java [hunk 9]
```diff
@@ -123,40 +149,10 @@ public Mono<StoreResponse> invokeAsync(RxDocumentServiceRequest request,
...
         };
 
         Function<Quadruple<Boolean, Boolean, Duration, Integer>, Mono<StoreResponse>> inBackoffFuncDelegate = null;
 
-        // we will enable fallback to other regions if the following conditions are met:
-        // 1. request is a read operation AND
-        // 2. enableReadRequestsFallback is set to true. (can only ever be true if
-        // direct mode, on client)
-        if (request.isReadOnlyRequest() && this.enableReadRequestsFallback) {
-            if (request.requestContext.cosmosDiagnostics == null) {
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java:136]: thanks for cleaning this part

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java [hunk 10]
```diff
@@ -171,6 +167,81 @@ public Mono<StoreResponse> invokeAsync(RxDocumentServiceRequest request,
...
     }
 
+    private Mono<StoreResponse> getStoreResponseMonoWithSpeculation(RxDocumentServiceRequest request, Quadruple<Boolean, Boolean, Duration, Integer> forceRefreshAndTimeout) {
+        CosmosE2EOperationRetryPolicyConfig config = request.requestContext.getE2EOperationRetryPolicyConfig();
+        List<Mono<StoreResponse>> monoList = new ArrayList<>();
+        List<RxDocumentServiceRequest> requests = new ArrayList<>();
+
+        if (speculativeProcessor.getRegionsForPureExploration() != null
+            && !speculativeProcessor.getRegionsForPureExploration().isEmpty()) {
+            explore(request, forceRefreshAndTimeout);
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java:177]: should the explore also be part of the Mono list

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java [hunk 11]
```diff
@@ -171,6 +167,81 @@ public Mono<StoreResponse> invokeAsync(RxDocumentServiceRequest request,
...
+                    );
+            }
+        }
+    }
+
+    private boolean shouldSpeculate(RxDocumentServiceRequest request) {
+        if (speculativeProcessor == null) {
+            return false;
+        }
+        if (!request.isReadOnlyRequest() && request.getResourceType() != ResourceType.Document) {
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java:232]: should the condition be `OR`?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java [hunk 12]
```diff
@@ -171,6 +167,81 @@ public Mono<StoreResponse> invokeAsync(RxDocumentServiceRequest request,
...
+            explore(request, forceRefreshAndTimeout);
+        }
+
+        if (speculativeProcessor.shouldIncludeOriginalRequestRegion()) {
+            monoList.add(getStoreResponseMono(request, forceRefreshAndTimeout));
+            requests.add(request);
+        }
+
+        for (URI endpoint : speculativeProcessor.getRegionsToSpeculate(config, this.transportClient
+            .getGlobalEndpointManager().getAvailableReadEndpoints())) {
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java:186]: same question here, should we use readEndpoints instead

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java (hunk 10 above)
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java:177]: can explore also be part of the Mono list ?I guess no because we do not want to cancel them (firstWithValue)

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java [hunk 13]
```diff
@@ -171,6 +167,81 @@ public Mono<StoreResponse> invokeAsync(RxDocumentServiceRequest request,
...
+
+            if (request.requestContext.locationEndpointToRoute != endpoint) {
+                RxDocumentServiceRequest newRequest = request.clone();
+                newRequest.requestContext.routeToLocation(endpoint);
+                requests.add(newRequest);
+                monoList.add(getStoreResponseMono(newRequest, forceRefreshAndTimeout).delaySubscription(speculativeProcessor.getThreshold(config)));
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java:192]: I am wondering how this will work for query for threshold based algo - the threshold is defined on operation, but for query, it will be per partition/feedRange request?  should the speculative processor be positioned in the RxDocumentClientImpl? (this is also where the e2e timeout is forced?)

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentServiceRequest.java [hunk 14]
```diff
@@ -1032,9 +1032,9 @@ public byte[] getContentAsByteArray() {
 
     @Override
     public RxDocumentServiceRequest clone() {
-        RxDocumentServiceRequest rxDocumentServiceRequest = RxDocumentServiceRequest.create(this.clientContext, this.getOperationType(), this.resourceId,this.getResourceType(),this.getHeaders());
+        RxDocumentServiceRequest rxDocumentServiceRequest = RxDocumentServiceRequest.create(this.clientContext, this.getOperationType(), this.resourceId, true, this.getResourceType(),this.getHeaders());
```
Comment (jeet1995) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentServiceRequest.java:1035]:
```suggestion
        RxDocumentServiceRequest rxDocumentServiceRequest = RxDocumentServiceRequest.create(this.clientContext, this.getOperationType(), this.resourceId, this.isNameBased, this.getResourceType(),this.getHeaders());
```

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentServiceRequest.java [hunk 15]
```diff
@@ -1048,9 +1048,18 @@ public RxDocumentServiceRequest clone() {
...
         rxDocumentServiceRequest.nonIdempotentWriteRetriesEnabled = this.nonIdempotentWriteRetriesEnabled;
+        rxDocumentServiceRequest.setResourceAddress(this.resourceAddress);
+        rxDocumentServiceRequest.requestContext = this.requestContext.clone();
+        rxDocumentServiceRequest.feedRange = this.feedRange;
+        rxDocumentServiceRequest.effectiveRange = this.effectiveRange;
         return rxDocumentServiceRequest;
     }
 
+    // Used by clone
+    private static RxDocumentServiceRequest create(DiagnosticsClientContext clientContext, OperationType operationType, String resourceId, boolean b, ResourceType resourceType, Map<String, String> headers) {
```
Comment (jeet1995) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/RxDocumentServiceRequest.java:1059]:
```suggestion
    private static RxDocumentServiceRequest create(DiagnosticsClientContext clientContext, OperationType operationType, String resourceId, boolean isNameBased, ResourceType resourceType, Map<String, String> headers) {
```

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java [hunk 16]
```diff
@@ -171,6 +167,81 @@ public Mono<StoreResponse> invokeAsync(RxDocumentServiceRequest request,
...
+                return storeResponse;
+            });
+    }
+
+    private void explore(RxDocumentServiceRequest request, Quadruple<Boolean, Boolean, Duration, Integer> forceRefreshAndTimeout) {
+        for (URI endpoint : speculativeProcessor.getRegionsForPureExploration()) {
+            if (request.requestContext.locationEndpointToRoute != endpoint) {
+                RxDocumentServiceRequest newRequest = request.clone();
+                newRequest.requestContext.routeToLocation(endpoint);
+                getStoreResponseMono(newRequest, forceRefreshAndTimeout)
```
Comment (jeet1995) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java:215]: Does it make sense to have a dedicated execution context / scheduler for exploration requests?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java [hunk 17]
```diff
@@ -171,6 +167,81 @@ public Mono<StoreResponse> invokeAsync(RxDocumentServiceRequest request,
...
+                    .subscribeOn(Schedulers.boundedElastic())
+                    .map(storeResponse ->
```
Comment (jeet1995) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java:217]: Does it make sense to have a dedicated execution context / scheduler for exploration requests?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/speculativeprocessors/ThompsonSamplingBasedSpeculation.java [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+
+    private final Map<URI, Double> alphaMap;
+    private final Map<URI, Double> betaMap;
+    private final Map<URI, Double> thetaMap;
+    private volatile Map<URI, Deque<Double>> rewardsMap;
+    private final URI localRegion;
+    private final EnumeratedDistribution<URI> armExplorationDist;
+    private final EnumeratedIntegerDistribution shouldExploreDist;
+
+
```
Comment (jeet1995) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/speculativeprocessors/ThompsonSamplingBasedSpeculation.java:39]: Nit: Extra line.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/speculativeprocessors/ThompsonSamplingBasedSpeculation.java [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+            if (uri.equals(localRegion)) {
+                continue;
+            }
+            pairs.add(new Pair<>(uri, 0.3 / (numRegions - 1))); // 30% of the time
+        }
+        return pairs;
+    }
+
+    public URI getSelection() {
+        for (URI endpoint : endpoints) {
```
Comment (jeet1995) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/speculativeprocessors/ThompsonSamplingBasedSpeculation.java:83]: `getSelection()` can be private

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/speculativeprocessors/ThompsonSamplingBasedSpeculation.java [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+        if (rewardsMap.get(uri).size() > DEFAULT_WINDOW_SIZE) {
+            double oldReward = rewardsMap.get(uri).pop();
+            alphaMap.computeIfPresent(uri, (k, v) -> v - oldReward);
+            betaMap.computeIfPresent(uri, (k, v) -> v - 1);
+        }
+        alphaMap.computeIfPresent(uri, (k, v) -> v + latencyInMs);
+        betaMap.computeIfPresent(uri, (k, v) -> v + 1);
+    }
+
+    public URI getRegionToExplore() {
```
Comment (jeet1995) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/speculativeprocessors/ThompsonSamplingBasedSpeculation.java:101]: Can be made private.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/speculativeprocessors/ThompsonSamplingBasedSpeculation.java [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+
+    public URI getSelection() {
+        for (URI endpoint : endpoints) {
+            thetaMap.put(endpoint, new GammaDistribution(alphaMap.get(endpoint), 1 / betaMap.get(endpoint)).sample());
+        }
+        return Collections.min(thetaMap.entrySet(), Map.Entry.comparingByValue()).getKey();
+    }
+
+    public void updateReward(URI uri, double latencyInMs) {
+        rewardsMap.get(uri).add(latencyInMs);
```
Comment (jeet1995) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/speculativeprocessors/ThompsonSamplingBasedSpeculation.java:90]: Can be made private.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosQueryRequestOptions.java [hunk 22]
```diff
@@ -827,8 +819,17 @@ public void applyMaxItemCount(
 
                     fluxOptions.setMaxItemCount(requestOptions.getMaxItemCount());
                 }
+
+                @Override
+                public CosmosE2EOperationRetryPolicyConfig getE2EOperationRetryConfig(CosmosQueryRequestOptions options) {
```
Comment (jeet1995) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosQueryRequestOptions.java:824]:
```suggestion
                public CosmosE2EOperationRetryPolicyConfig getE2EOperationRetryPolicyConfig(CosmosQueryRequestOptions options) {
```

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java (hunk 13 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java:192]: Also we would not want to speculate on all regions at the same time I guess? Wouldn't it make more sense to start after x ms on seocnd region, then aftera nother x ms on third region or similar?

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java:192]: But yeah - for now I can also live with the current model

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/HttpConstants.java (hunk 4 above)
Comment (mbhaskar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/HttpConstants.java:378]: Oops this is stale code as I branched this off of my earlier PR. Will remove

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java (hunk 6 above)
Comment (mbhaskar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java:33]: On 0 - no speculation, that is just for sanity. Any value other than 1 or 2 will essentially mean no speculation.

On the first question, yes until we get to an optimal implementation we can test both and see how that goes

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java (hunk 7 above)
Comment (mbhaskar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java:105]: Thanks for the suggestion. Will use readEndpoints

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java (hunk 13 above)
Comment (mbhaskar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java:192]: Not very difficult to do. Can add another config parameter STEP that makes the speculation incremental

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java (hunk 11 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java:232]: ah gotcha. so we only target for reads in this PR

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java (hunk 13 above)
Comment (mbhaskar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java]: Added threshold step

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java (hunk 17 above)
Comment (mbhaskar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java]: removed exploration as only threshold based speculation is in

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/speculativeprocessors/ThompsonSamplingBasedSpeculation.java (hunk 20 above)
Comment (mbhaskar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/speculativeprocessors/ThompsonSamplingBasedSpeculation.java]: Removed this class for now

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/speculativeprocessors/ThompsonSamplingBasedSpeculation.java [hunk 23]
```diff
@@ -0,0 +1,137 @@
...
+        betaMap.computeIfPresent(uri, (k, v) -> v + 1);
+    }
+
+    public URI getRegionToExplore() {
+        // Should return the next arm to explore. This is random with max weight on local
+        return armExplorationDist.sample();
+    }
+
+    public boolean shouldExplore() {
+        return shouldExploreDist.sample() == 1;
```
[the remark this answers is not in the document]
Comment (mbhaskar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/speculativeprocessors/ThompsonSamplingBasedSpeculation.java]: Removed this class for now

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/speculativeprocessors/ThompsonSamplingBasedSpeculation.java (hunk 19 above)
Comment (mbhaskar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/speculativeprocessors/ThompsonSamplingBasedSpeculation.java]: Removed this class for now

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosDiagnostics.java [hunk 24]
```diff
@@ -33,7 +33,7 @@
  */
 public final class CosmosDiagnostics {
     private static final Logger LOGGER = LoggerFactory.getLogger(CosmosDiagnostics.class);
-    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+    static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosDiagnostics.java:36]: why not private anymore? Are we re-using this anywhere else?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java [hunk 25]
```diff
@@ -300,6 +311,18 @@ public static int getSessionTokenMismatchMaximumBackoffTimeInMs() {
             DEFAULT_SESSION_TOKEN_MISMATCH_MAXIMUM_BACKOFF_TIME_IN_MILLISECONDS);
     }
 
+    public static int getSpeculationType() {
+        return getJVMConfigAsInt(SPECULATION_TYPE, 0);
+    }
+
+    public static int speculationThreshold() {
+        return getJVMConfigAsInt(SPECULATION_THRESHOLD, 500);
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java:319]: nit: we should define these default values in separate constant variables above like other methods.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/speculativeprocessors/ThresholdBasedSpeculation.java [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    @Override
+    public boolean shouldIncludeOriginalRequestRegion() {
+        return true;
+    }
+
+    @Override
+    public void onResponseReceived(URI region, Duration latency) {
+        // unused
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/speculativeprocessors/ThresholdBasedSpeculation.java:40]: if it is unused, may be throw error if it gets called, something like not implemented exception. Can be done in a follow up PR.

File: sdk/cosmos/azure-cosmos/src/main/java/module-info.java [hunk 27]
```diff
@@ -28,7 +28,7 @@
     requires java.logging;
 	requires HdrHistogram;
 
-	// public API surface area
+    // public API surface area
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/src/main/java/module-info.java:31]: can revert this change,.

**Review by kushagraThapar — APPROVED** [review]
LGTM, thanks @mbhaskar

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java [hunk 28]
```diff
@@ -300,6 +311,18 @@ public static int getSessionTokenMismatchMaximumBackoffTimeInMs() {
             DEFAULT_SESSION_TOKEN_MISMATCH_MAXIMUM_BACKOFF_TIME_IN_MILLISECONDS);
     }
 
+    public static int getSpeculationType() {
+        return getJVMConfigAsInt(SPECULATION_TYPE, 0);
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java:315]: should this by default be 1?

**Review by xinlian12 — APPROVED** [review]
LGTM, thanks

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/DocumentServiceRequestContext.java [hunk 29]
```diff
@@ -127,7 +127,6 @@ public DocumentServiceRequestContext clone() {
         context.throughputControlCycleId = this.throughputControlCycleId;
         context.replicaAddressValidationEnabled = this.replicaAddressValidationEnabled;
         context.e2EOperationRetryPolicyConfig = this.e2EOperationRetryPolicyConfig;
-        context.effectivePartitionKey = this.effectivePartitionKey;
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/DocumentServiceRequestContext.java]: why was this removed?

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/DocumentServiceRequestContext.java]: Blocking

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/DocumentServiceRequestContext.java]: Never mind - just a change to the baseline I last reviewed.

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM - Thanks

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosDiagnostics.java (hunk 24 above)
Comment (mbhaskar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosDiagnostics.java:36]: Was using in tests to validate.

Comment (mbhaskar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosDiagnostics.java:36]: But could revert.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java (hunk 28 above)
Comment (mbhaskar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/Configs.java:315]: No speculation by default

File: sdk/cosmos/azure-cosmos/src/main/java/module-info.java (hunk 27 above)
Comment (mbhaskar) [reply sdk/cosmos/azure-cosmos/src/main/java/module-info.java:31]: Will revert in follow up PR
